src/lib/Server/Output/ValueObjectVisitor/Location.php

```diff
@@ -174,4 +174,27 @@ protected function visitLocationAttributes(Visitor $visitor, Generator $generato
 
         $generator->endObjectElement('ContentInfo');
     }
+
+    /**
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     */
+    private function resolveMainLocation(
```

Comment:
Does it make sense to add a logger here to track content without main locations? There always should be one main location.

Reply:
Definitely makes sense, I'll add one.

Reply:
Resolved in https://github.com/ezsystems/ezplatform-rest/pull/100/commits/383a8dfa7c741955ec19bdd557be3596435f12e5

Reply:
@lserwatka @barw4 
I am not sure if that exactly true, even our phpdocs states that `null` is allowed  in `mainLocationID`, and it is the case for drafts, right? Now we gonna log those, and that could be confusing.

@ezsystems/php-dev-team

Reply:
Yes, no main location is perfectly valid from API point of view. It happens when you fetch content that is not yet published. I'm also not convinced this should be logged in any way.

Reply:
We are providing a type hinted `Location` to `visit()` though so we assume that a content is not a draft.

Reply:
Actually, there was a situation where a content was trashed and was still indexed in Solr, logging would help to find such a problematic content.

Reply:
Content can have no Locations at all regardless if it's a first draft or published version.

Reply:
That make sense then, I would just consider logging it into separate channel maybe?

Reply:
Yes, @barw4 has a use case where we can benefit from logging

Reply:
Both channel and log level should be picked correctly by us.

```php
    /**
     * Exceptional occurrences that are not errors.
     *
     * Example: Use of deprecated APIs, poor use of an API, undesirable things
     * that are not necessarily wrong.
     *
     * @param string $message
     * @param array $context
     * @return void
     */
    public function warning($message, array $context = array());

    /**
     * Normal but significant events.
     *
     * @param string $message
     * @param array $context
     * @return void
     */
    public function notice($message, array $context = array());

    /**
     * Interesting events.
     *
     * Example: User logs in, SQL logs.
     *
     * @param string $message
     * @param array $context
     * @return void
     */
    public function info($message, array $context = array());
```
(https://www.php-fig.org/psr/psr-3/)

In this case, if we can differentiate between states (if we expect a `null` main location, or is it actually an error) it would be ideal to use different logging levels. Most production systems are configured to not log `info` level messages, and oftentimes warnings are also skipped (usually with fingers crossed and no actual error).

If we're going to start using channels, we should document them.

Reply:
@ezsystems/php-dev-team  So, what's the consensus on logging?

Obviously, API allows for content to have a `null` main location, however, I would say having no main location in this particular case, when visiting other location of this particular content (we have to provide a `Location` object), seems like something that is abnormal, because we are considering the situation when content has location without having main location.

Reply:
> In this case, if we can differentiate between states (if we expect a null main location, or is it actually an error) it would be ideal to use different logging levels. Most production systems are configured to not log info level messages, and oftentimes warnings are also skipped (usually with fingers crossed and no actual error).

> If we're going to start using channels, we should document them.

@Steveb-p given @barw4 feedback I think it could be warning in this case. AFAIR we already have `ibexa` channel. Or should it be a different one and if so how should we name it? Seems this is broader topic.

Reply:
This was exactly my point. When you have locations, one location has to be 'main' otherwise we have abnormal situation. I'm not sold on the idea we need dedicated channel, this is clear indicator something was corrupted or API was used in a wrong way. @webhdx could you make a final call based on internal team discussion?

Reply:
@lserwatka I agree this is abnormal situation, but I do not think REST Location visitor is a proper place to check and log it. If we agree this situation may happen (it won't most likely be due to faulty API but may be the result of failed migrations and such), we should check and log it much earlier, on DomainMapper level I guess.

Reply:
Good point, I'm not against of adding this logger in a better place. This will make debugging life easier.

Reply:
@barw4 
If we log this in DomainMapper, is there a chance that assigning one of the locations as a mainLocation there (if it is empty, because of some bug) would prevent us to seeing that issue in the future?

Reply:
@ezsystems/php-dev-team please, see https://github.com/ezsystems/ezplatform-kernel/pull/364